src/js/fable-core/Array2.fs

```diff
@@ -1,16 +1,750 @@
+    for arr in arrays do
+        totalLength <- totalLength + arr.Length
+
+    let result = cons.Create totalLength
+    let mutable offset = 0;
+    for arr in arrays do
+        typedArraySetImpl result arr <| float(offset)
+        offset <- offset + arr.Length
+    result
+
+let private concatImpl (cons: ArrayCons) (arrays: 'T[][]): 'T[] =
+    if arrays.Length > 0 then
+        if isTypedArrayImpl arrays.[0] then
+            typedArrayConcatImpl cons arrays
+        else
+            dynamicArrayConcatImpl arrays
+    else
+        cons.Create 0
+
+let concat (arrays: 'T[] seq) (cons: ArrayCons): 'T[] =
+    arrays
+    |> Seq.toArray
+    |> concatImpl cons
+
+let collect (mapping: 'T -> 'U[]) (array: 'T[]) (cons: ArrayCons): 'U[] =
+    map mapping array cons
+    |> concatImpl cons
+
+let countBy (projection: 'T->'Key) (array: 'T[]) (cons: ArrayCons) =
+    let dict = System.Collections.Generic.Dictionary<'Key, int>()
+
+    for value in array do
+        let key = projection value
+        let mutable prev = Unchecked.defaultof<_>
+        if dict.TryGetValue(key, &prev) then dict.[key] <- prev + 1 else dict.[key] <- 1
+
+    let res = cons.Create dict.Count
+    let mutable i = 0
+    for group in dict do
+        res.[i] <- group.Key, group.Value
+        i <- i + 1
+    res
+
+let distinctBy projection (array:'T[]) (cons: ArrayCons) =
+    let temp = cons.Create array.Length
+    let mutable i = 0
+
+    let hashSet = System.Collections.Generic.HashSet<'T>()
+    for v in array do
+        if hashSet.Add(projection v) then
+            temp.[i] <- v
+            i <- i + 1
+
+    sliceImpl temp 0 i
+
+let distinct (array: 'T[]) (cons: ArrayCons) = distinctBy id array cons
+
+let where predicate (array: _[]) = filterImpl predicate array
+
+let except (itemsToExclude: seq<_>) (array:_[]) =
+    if array.Length = 0 then
+        array
+    else
+        let cached = System.Collections.Generic.HashSet(itemsToExclude)
+        array |> filterImpl cached.Add
+
+let groupBy (projection: 'T->'Key) (array: 'T[]) (cons: ArrayCons) =
+    let dict = System.Collections.Generic.Dictionary<'Key, 'T[]>()
+
+    // Build the groupings
+    for i = 0 to (array.Length - 1) do
+        let v = array.[i]
+        let key = projection v
+        let mutable prev = Unchecked.defaultof<_>
+        if dict.TryGetValue(key, &prev) then
+            pushImpl prev v |> ignore
+        else
+            // Use dynamic array so we can build it up via .push()
+            // Consider benchmarking if another collection type performs better here
+            let prev = [|v|]
+            dict.[key] <- prev
+
+    // Return the array-of-arrays.
+    let result = cons.Create dict.Count
+    let mutable i = 0
+    for group in dict do
+        result.[i] <- group.Key, cons.From group.Value
+        i <- i + 1
+
+    result
+
+let ofArray (arr: 'T[]) = arr
+
+let inline private emptyImpl (cons: ArrayCons) = cons.Create(0)
+
+let empty = emptyImpl
```

Comment:
So, `empty` expects `cons` now, and `skip`, `skipWhile`, `take` and `takeWhile` also do as depending on `empty`.